// graphql-schema-decorator/lib/pagination.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql = require("graphql");
const page_info_type_1 = require("./page-info.type");
class PaginationResponse {
    constructor(count, nodes, pageInfo) {
        this.count = count;
        this.nodes = nodes;
        this.pageInfo = pageInfo;
    }
}
exports.PaginationResponse = PaginationResponse;
class PaginationRequest {
    constructor(offset, limit, orderBy) {
        this.offset = offset;
        this.limit = limit;
        this.orderBy = orderBy;
    }
}
exports.PaginationRequest = PaginationRequest;
class PaginationType {
    static build(name, type) {
        return new graphql.GraphQLObjectType({
            name: `${name}Connection`,
            description: `Connecton object that connects '${name}' to the server`,
            fields: function () {
                return {
                    count: {
                        type: graphql.GraphQLInt,
                        description: 'Total number of elements',
                    },
                    nodes: {
                        type: new graphql.GraphQLList(type),
                        description: 'Array of elements',
                    },
                    pageInfo: {
                        type: page_info_type_1.PageInfoType,
                        description: 'Pagination information',
                    },
                };
            },
        });
    }
}
exports.PaginationType = PaginationType;
